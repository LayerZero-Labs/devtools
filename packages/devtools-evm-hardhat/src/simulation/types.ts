import { AnvilOptions } from '@layerzerolabs/devtools-evm'

export interface SimulationUserConfig {
    /**
     * Simulation works by creating local forks of networks specified
     * in hardhat config. These networks are containerized and not accessible
     * from the developer machine - instead, a simple proxy server
     * listening on this port proxies requests to the individual networks.
     *
     * For example, a hardhat configuration with networks `fuji` and `mainnet`
     * will result in two forks being created. These networks will then be accessible
     * on `http://localhost:<port>/fuji` and `http://localhost:<port>/mainnet`
     *
     * @default 8545
     */
    port?: number

    /**
     * Simulation task stores its artifacts on the local filesystem.
     *
     * To customize the path, set this property to a relative or absolute path
     * (relative path will be resolved against the root path of your hardhat project)
     *
     * @default .layerzero
     */
    directory?: string

    /**
     * Simulation will by default overwrite the network accounts by values
     * provided in Anvil config (or defaults if not present).
     *
     * The mnemonic will default to "test test test test test test test test test test test junk"
     * and the number of funded accounts to 10.
     *
     * @default true
     */
    overwriteAccounts?: boolean

    /**
     * Anvil overrides for the underlying EVM nodes.
     */
    anvil?: SimulationAnvilUserConfig
}

/**
 * Resolved simulation config
 */
export interface SimulationConfig {
    port: number
    directory: string
    overwriteAccounts: boolean
    anvil: SimulationAnvilConfig
}

/**
 * Resolved simulation config for anvil.
 *
 * This config is created by taking the user config
 * and applying defaults.
 */
export type SimulationAnvilConfig = AnvilOptions & {
    mnemonic: NonNullable<AnvilOptions['mnemonic']>
}

/**
 * User facing simulation config for anvil.
 *
 * This config cannot override several system attributes for anvil
 * (since it would not really make sense seeing that anvil is being run in containers):
 *
 * - host & port: this would break the docker port bindings
 * - state: state is kept within the container and is not accessible from the outside
 * - forkUrl: this is set based on the hardhat config and points to the hardhat network url
 */
export type SimulationAnvilUserConfig = Omit<AnvilOptions, 'host' | 'port' | 'state' | 'forkUrl'>
