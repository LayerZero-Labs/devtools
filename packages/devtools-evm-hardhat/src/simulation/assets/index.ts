// Even though it would be nice to just call this file Dockerfile instead of Dockerfile.conf,
// esbuild (or tsup) have issues with specifying loaders for files without extensions.
//
// And since we already have a file with .conf extension,
// we add the same extension to the Dockerfile to use the same d.ts file and the same loader
export { default as dockerfile } from './Dockerfile.conf'
export { default as nginxConf } from './nginx.conf'
