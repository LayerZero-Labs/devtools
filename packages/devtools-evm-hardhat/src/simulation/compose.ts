import { pipe } from 'fp-ts/lib/function'
import * as RR from 'fp-ts/ReadonlyRecord'
import type { ComposeSpec, ComposeSpecService } from '@layerzerolabs/devtools'
import { type AnvilOptions, createAnvilCliOptions } from '@layerzerolabs/devtools-evm'
import type { ComposeSpecServices } from '@layerzerolabs/devtools'
import type { SimulationConfig } from './types'

/**
 * Creates a docker compose service specification for an anvil-based EVM node
 *
 * @param {AnvilOptions} anvilOptions
 * @returns {ComposeSpecService}
 */
export const createEvmNodeServiceSpec = (anvilOptions: AnvilOptions): ComposeSpecService => ({
    // This service references a Dockerfile that is copied
    // next to the resulting docker-compose.yaml
    //
    // The source for this Dockerfile is located in src/simulation/assets/Dockerfile.conf
    build: {
        dockerfile: 'Dockerfile',
        target: 'node-evm',
    },
    command: ['anvil', ...createAnvilCliOptions(anvilOptions)],
})

/**
 * Creates a docker compose service specification for an nginx-based proxy service
 * that proxies requests to underlying EVM nodes (or their RPC URLs to be mor precise)
 *
 * @param {number} port
 * @param {ComposeSpecServices} networkServices
 * @returns {ComposeSpecService}
 */
export const createEvmNodeProxyServiceSpec = (
    port: number,
    networkServices: ComposeSpecServices
): ComposeSpecService => ({
    // This service references a Dockerfile that is copied
    // next to the resulting docker-compose.yaml
    //
    // The source for this Dockerfile is located in src/simulation/assets/Dockerfile.conf
    build: {
        dockerfile: 'Dockerfile',
        target: 'proxy-evm',
    },
    // This service will expose its internal 8545 port to a host port
    //
    // The internal 8545 port is hardcoded both here and in the nginx.conf file,
    // the source for which is located in src/simulation/assets/nginx.conf
    ports: [`${port}:8545`],
    depends_on: pipe(
        networkServices,
        // This service will depend on the RPCs to be healthy
        // so we'll take the networkServices object and replace
        // the values with service_healthy condition
        RR.map(() => ({
            condition: 'service_healthy',
        }))
    ),
})

/**
 * Creates a docker compose spec with a set of anvil-based EVM nodes
 * and a single proxy server that proxies requests to these nodes.
 *
 * @param {SimulationConfig} config
 * @param {Record<string, AnvilOptions>} networks
 * @returns {ComposeSpec}
 */
export const createSimulationComposeSpec = (
    config: SimulationConfig,
    networks: Record<string, AnvilOptions>
): ComposeSpec => ({
    version: '3.9',
    services: pipe(
        networks,
        // First we turn the networks into docker compose specs for EVM nodes
        RR.map(createEvmNodeServiceSpec),
        (networkServiceSpecs) =>
            // Then we add the RPC proxy server
            //
            // There is a small edge case here that we can address
            // if it ever comes up: if a network is called 'rpc', this compose file
            // will not work.
            //
            // The fix for this is to prefix all networks with something like network-xxx
            // but we can do that if ever this usecase comes up
            pipe(
                networkServiceSpecs,
                RR.upsertAt('rpc', createEvmNodeProxyServiceSpec(config.port, networkServiceSpecs))
            )
    ),
})
