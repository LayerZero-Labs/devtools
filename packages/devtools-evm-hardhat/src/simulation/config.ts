import type { HardhatConfig, HttpNetworkConfig, NetworkConfig } from 'hardhat/types'
import type { SimulationConfig, SimulationUserConfig } from './types'
import { resolve } from 'path'
import { AnvilOptions } from '@layerzerolabs/devtools-evm'
import { pipe } from 'fp-ts/lib/function'
import * as R from 'fp-ts/Record'

/**
 * Turns `SimulationUserConfig` into `SimulationConfig` by supplying defaults
 *
 * @param {SimulationUserConfig} userConfig
 * @param {HardhatConfig} hardhatConfig
 * @returns {SimulationConfig}
 */
export const resolveSimulationConfig = (
    userConfig: SimulationUserConfig,
    hardhatConfig: HardhatConfig
): SimulationConfig => ({
    port: userConfig.port ?? 8545,
    directory: resolve(hardhatConfig.paths.root, userConfig.directory ?? '.layerzero'),
    overwriteAccounts: userConfig.overwriteAccounts ?? true,
    anvil: {
        // For now we'll hardcode the mnemonic we'll use to seed the accounts on the simulation networks
        mnemonic: 'test test test test test test test test test test test junk',
        ...userConfig.anvil,
        // The host and port need to always point to 0.0.0.0:8545
        // since anvil runs in the container that exposes this port on 0.0.0.0
        host: '0.0.0.0',
        port: 8545,
    },
})

/**
 * Takes a portion of the hardhat networks config and turns the values into `AnvilOptions`
 * to be used by a simulation network container.
 *
 * This is used in the simulation generation phase where the compose files
 * for networks defined in hardhat config are being generated.
 *
 * The `Record<string, NetworkConfig>` is used instead of `NetworksConfig` type from hardhat
 * to avoid any type issues with special keys like `localhost` or `hardhat`.
 *
 * @param {SimulationConfig} config
 * @param {Record<string, NetworkConfig>} networksConfig
 * @returns {Record<string, AnvilOptions>} Object with the same keys as the networks config, with values mapped to `AnvilOptions`
 */
export const getAnvilOptionsFromHardhatNetworks = (
    config: SimulationConfig,
    networksConfig: Record<string, NetworkConfig>
): Record<string, AnvilOptions> =>
    pipe(
        networksConfig,
        // We want to drop all the networks that don't have URLs
        R.filter(isHttpNetworkConfig),
        // And map the network configs into AnvilOptions
        R.map(
            (networkConfig: HttpNetworkConfig): AnvilOptions => ({
                ...config.anvil,
                forkUrl: networkConfig.url,
            })
        )
    )

/**
 * Returns with overrides for hardhat network configuration.
 *
 * This, when merged with a hardhat config, will redirect the network calls to the local EVM nodes.
 *
 * @param {SimulationConfig} config
 * @param {Record<string, NetworkConfig>} networksConfig
 * @returns {Record<string, NetworkConfig>}
 */
export const getHardhatNetworkOverrides = (
    config: SimulationConfig,
    networksConfig: Record<string, NetworkConfig>
): Record<string, NetworkConfig> =>
    pipe(
        networksConfig,
        // We want to drop all the networks that don't have URLs
        R.filter(isHttpNetworkConfig),
        // We'll take the existing network configs and point them to our RPC proxy
        //
        // It's important that these configs are not saved to filesystem as they might contain
        // sensitive data (and forgetting to ignore these files in git could lead to security breaches)
        R.mapWithIndex(
            (networkName, networkConfig): NetworkConfig => ({
                ...networkConfig,
                // We want to redirect this network to the local proxy
                //
                // This is the nginx server listening on the port we configured in the simulation configuration
                url: new URL(networkName, `http://localhost:${config.port}`).toString(),
                accounts: config.overwriteAccounts
                    ? // When overwriting accounts, all the accounts will be switched to the anvil config
                      // (or reasonable defaults if not provided)
                      {
                          mnemonic: config.anvil.mnemonic,
                          // These need to be defaulted to the anvil options
                          // (or the anvil defaults)
                          //
                          // See https://book.getfoundry.sh/reference/cli/anvil for anvil defaults
                          count: config.anvil.count ?? 10,
                          path: config.anvil.derivationPath ?? "m/44'/60'/0'/0/",
                          // These will be hardcoded for now as anvil does not support setting these
                          initialIndex: 0,
                          passphrase: '',
                      }
                    : networkConfig.accounts,
            })
        )
    )

/**
 * Helper utility to pick network configs by their names / object keys.
 *
 * Similar to TypeScript `Pick` helper type, but in runtime.
 *
 * @param {string[]} networks List of networks to pick
 */
export const pickNetworkConfigs = (networks: string[]) =>
    R.filterWithIndex<string, NetworkConfig>((networkName: string) => networks.includes(networkName))

/**
 * Little helper utility that checks whether a network config is not a hardhat network config
 *
 * @param {NetworkConfig} networkConfig
 * @returns {boolean}
 */
const isHttpNetworkConfig = (networkConfig: NetworkConfig): networkConfig is HttpNetworkConfig =>
    'url' in networkConfig && typeof networkConfig.url === 'string'
