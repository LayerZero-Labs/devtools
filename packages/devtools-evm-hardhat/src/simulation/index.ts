export * from './compose'
export * from './config'
export * from './types'
