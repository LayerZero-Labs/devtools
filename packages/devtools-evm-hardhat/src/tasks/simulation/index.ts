export * from './logs'
export * from './start'
export * from './stop'
