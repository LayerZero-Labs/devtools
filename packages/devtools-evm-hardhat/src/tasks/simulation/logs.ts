import '@/type-extensions'

import { TASK_LZ_TEST_SIMULATION_LOGS, TASK_LZ_TEST_SIMULATION_START } from '@/constants'
import { ActionType } from 'hardhat/types'
import { task } from 'hardhat/config'
import { printLogo } from '@layerzerolabs/io-devtools/swag'
import { LogLevel, createLogger, isFile, printJson, setDefaultLogLevel } from '@layerzerolabs/io-devtools'
import { types } from '@/cli'
import { resolveSimulationConfig } from '@/simulation/config'
import { join } from 'path'
import { spawnSync } from 'child_process'

export interface SimulationLogsTaskArgs {
    logLevel?: LogLevel
}

const action: ActionType<SimulationLogsTaskArgs> = async ({ logLevel = 'info' }, hre) => {
    setDefaultLogLevel(logLevel)

    printLogo()

    const logger = createLogger()

    // Grab the simulation user config from hardhat user config
    const simulationUserConfig = hre.userConfig.layerZero?.experimental?.simulation ?? {}
    logger.verbose(`Using simulation user config:\n${printJson(simulationUserConfig)}`)

    // Resolve the defaults for the simulation config
    const simulationConfig = resolveSimulationConfig(simulationUserConfig, hre.config)
    logger.verbose(`Resolved simulation config:\n${printJson(simulationConfig)}`)

    // Check that the docker compose file exists
    const dockerComposePath = join(simulationConfig.directory, 'docker-compose.yaml')
    if (!isFile(dockerComposePath)) {
        logger.warn(`Could not find simulation docker compose file '${dockerComposePath}'`)
        logger.warn(`Did you run 'npx hardhat ${TASK_LZ_TEST_SIMULATION_START}'?`)

        process.exitCode = 1

        return
    }

    // Spawn docker compose logs command, piping the stdout and stderr to the current shell
    //
    // The error reporting on this part should be improved - we should check that "docker" and "docker compose"
    // are known commands before we go ahead and try executing them
    try {
        logger.verbose(`Spawning docker compose logs command for ${dockerComposePath}`)

        spawnSync('docker', ['compose', '-f', dockerComposePath, 'logs'], {
            stdio: 'inherit',
        })
    } catch (error) {
        logger.error(`Failed to spawn docker compose logs command for ${dockerComposePath}: ${error}`)

        process.exitCode = 1

        return
    }
}

if (process.env.LZ_ENABLE_EXPERIMENTAL_SIMULATION) {
    task(TASK_LZ_TEST_SIMULATION_LOGS, 'Show logs for LayerZero omnichain simulation', action).addParam(
        'logLevel',
        'Logging level. One of: error, warn, info, verbose, debug, silly',
        'info',
        types.logLevel
    )
}
