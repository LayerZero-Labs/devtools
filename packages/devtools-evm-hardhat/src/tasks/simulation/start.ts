import '@/type-extensions'

import { TASK_LZ_TEST_SIMULATION_LOGS, TASK_LZ_TEST_SIMULATION_START } from '@/constants'
import { ActionType } from 'hardhat/types'
import { task } from 'hardhat/config'
import { types } from '@/cli'
import { printLogo } from '@layerzerolabs/io-devtools/swag'
import { LogLevel, createLogger, printJson, setDefaultLogLevel } from '@layerzerolabs/io-devtools'
import { mkdirSync, writeFileSync } from 'fs'
import { join } from 'path'
import { serializeDockerComposeSpec } from '@layerzerolabs/devtools'
import { createSimulationComposeSpec } from '@/simulation/compose'
import { assertDefinedNetworks } from '@/internal/assertions'
import { getEidsByNetworkName } from '@/runtime'
import { getAnvilOptionsFromHardhatNetworks, pickNetworkConfigs, resolveSimulationConfig } from '@/simulation'
import { dockerfile, nginxConf } from '@/simulation/assets'
import { spawnSync } from 'child_process'
import { EndpointId, Stage, endpointIdToStage } from '@layerzerolabs/lz-definitions'

export interface SimulationStartTaskArgs {
    logLevel?: LogLevel
    networks?: string[]
    daemon?: boolean
    stage?: Stage
}

const action: ActionType<SimulationStartTaskArgs> = async (
    { networks: networksArgument, daemon = false, logLevel = 'info', stage },
    hre
) => {
    setDefaultLogLevel(logLevel)

    printLogo()

    const logger = createLogger()

    // --stage cannot be used in conjunction with --networks
    if (networksArgument != null && stage != null) {
        logger.error(`--stage ${stage} cannot be used in conjunction with --networks ${networksArgument.join(',')}`)

        process.exit(1)
    }

    // And we create a filtering predicate for the stage argument
    const isOnStage = stage == null ? () => true : (eid: EndpointId) => endpointIdToStage(eid) === stage

    // Let's grab the networks that will be included in the simulation
    const networks = networksArgument
        ? // Here we need to check whether the networks have been defined in hardhat config
          assertDefinedNetworks(networksArgument)
        : //  But here we are taking them from hardhat config so no assertion is necessary
          Object.entries(getEidsByNetworkName()).flatMap(([networkName, eid]) =>
              eid != null && isOnStage(eid) ? [networkName] : []
          )

    // We only continue if we have any networks with eid
    //
    // The eid is not really a requirement - we could spin up forked nodes for any
    // network defined in the hardhat config - it's just a way of limiting this simulation to LayerZero networks
    if (networks.length === 0) {
        logger.warn(`No networks with eid configured, exiting`)

        return
    }

    logger.info(`Will create a simulation configuration for networks ${networks.join(', ')}`)

    // Grab the simulation user config from hardhat user config
    const simulationUserConfig = hre.userConfig.layerZero?.experimental?.simulation ?? {}
    logger.verbose(`Using simulation user config:\n${printJson(simulationUserConfig)}`)

    // Resolve the defaults for the simulation config
    const simulationConfig = resolveSimulationConfig(simulationUserConfig, hre.config)
    logger.verbose(`Resolved simulation config:\n${printJson(simulationConfig)}`)

    // Grab only the network configs we are going to need for the simulation
    const networkConfigs = pickNetworkConfigs(networks)(hre.config.networks)

    // Turn the network configs into anvil options
    const anvilOptions = getAnvilOptionsFromHardhatNetworks(simulationConfig, networkConfigs)
    logger.verbose(`The anvil config is:\n${printJson(anvilOptions)}`)

    // Now create the compose file
    const composeSpec = createSimulationComposeSpec(simulationConfig, anvilOptions)
    const serializedComposeSpec = serializeDockerComposeSpec(composeSpec)

    // And finally we write all the required file artifacts to filesystem
    logger.verbose(`Making sure directory ${simulationConfig.directory} exists`)
    mkdirSync(simulationConfig.directory, { recursive: true })

    const dockerfilePath = join(simulationConfig.directory, 'Dockerfile')
    logger.debug(`Writing simulation Dockerfile to ${dockerfilePath}`)
    writeFileSync(dockerfilePath, dockerfile)

    const nginxConfPath = join(simulationConfig.directory, 'nginx.conf')
    logger.debug(`Writing simulation nginx configuration file to ${nginxConfPath}`)
    writeFileSync(nginxConfPath, nginxConf)

    const dockerComposePath = join(simulationConfig.directory, 'docker-compose.yaml')
    logger.debug(`Writing simulation docker compose spec file to ${dockerComposePath}`)
    writeFileSync(dockerComposePath, serializedComposeSpec)

    // Spawn docker compose up command, piping the stdout and stderr to the current shell
    //
    // The error reporting on this part should be improved - we should check that "docker" and "docker compose"
    // are known commands before we go ahead and try executing them
    try {
        if (daemon) {
            logger.info(`Starting simulation in the background`)
            logger.info(
                `Use 'LZ_ENABLE_EXPERIMENTAL_SIMULATION=1 npx hardhat ${TASK_LZ_TEST_SIMULATION_LOGS}' to view the network logs`
            )
        } else {
            logger.info(`Starting simulation`)
        }

        logger.verbose(`Spawning docker compose up command for ${dockerComposePath}`)

        // This is a very quick and dirty way to pass an optional --wait argument to docker compose up
        const additionalUpArgs: string[] = daemon ? ['--wait'] : []

        spawnSync('docker', ['compose', '-f', dockerComposePath, 'up', ...additionalUpArgs], {
            stdio: 'inherit',
        })
    } catch (error) {
        logger.error(`Failed to spawn docker compose up command for ${dockerComposePath}: ${error}`)

        process.exitCode = 1

        return
    }
}

if (process.env.LZ_ENABLE_EXPERIMENTAL_SIMULATION) {
    task(TASK_LZ_TEST_SIMULATION_START, 'Start LayzerZero omnichain simulation', action)
        .addParam('logLevel', 'Logging level. One of: error, warn, info, verbose, debug, silly', 'info', types.logLevel)
        .addParam('networks', 'Comma-separated list of networks to simulate', undefined, types.csv, true)
        .addParam('stage', 'Chain stage. One of: mainnet, testnet, sandbox', undefined, types.stage, true)
        .addFlag('daemon', 'Start the simulation in the background')
}
