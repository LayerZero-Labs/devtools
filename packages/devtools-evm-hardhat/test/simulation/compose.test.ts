import hre from 'hardhat'
import { resolveSimulationConfig } from '@/simulation'
import {
    createEvmNodeProxyServiceSpec,
    createEvmNodeServiceSpec,
    createSimulationComposeSpec,
} from '@/simulation/compose'
import { serializeDockerComposeSpec } from '@layerzerolabs/devtools'
import { AnvilOptions } from '@layerzerolabs/devtools-evm'
import { mnemonicArbitrary, optionalArbitrary } from '@layerzerolabs/test-devtools'
import { spawnSync } from 'child_process'
import fc from 'fast-check'
import { rm, writeFile } from 'fs/promises'
import { join } from 'path'

describe('simulation/compose', () => {
    // The available docker port range
    const portArbitrary = fc.integer({ min: 1, max: 65535 })

    const anvilOptionsArbitrary: fc.Arbitrary<AnvilOptions> = fc.record({
        host: optionalArbitrary(fc.ipV4()),
        port: optionalArbitrary(portArbitrary),
        count: optionalArbitrary(fc.integer()),
        mnemonic: optionalArbitrary(mnemonicArbitrary),
        blockTime: optionalArbitrary(fc.integer()),
        forkUrl: optionalArbitrary(fc.webUrl()),
    })

    // We are aware of shortcomings of the simulation spec generation process:
    //
    // - whitespace service names will generate invalid compose files. This should not be a problem in a real hardhat project
    // - network named rpc will collide with the RPC proxy container defined in the spec
    //
    // Because of this we'll filter down on the possibilities when it comes to service names
    const serviceNameArbitrary = fc.constantFrom('mumbai', 'sepolia-testnet', 'someNetwork', 'someNetwork.V2')

    const SPEC_FILE_PATH = join(__dirname, 'docker-compose.yaml')

    // We add --log-level to suppress any warnings - for example warnings about environment variables not being defined
    //
    // This allows us to easily test the stderr for being empty (otherwise we would need to test it for not containing errors)
    const validateSpec = () => spawnSync('docker', ['--log-level', 'ERROR', 'compose', '-f', SPEC_FILE_PATH, 'config'])

    afterEach(async () => {
        await rm(SPEC_FILE_PATH, { force: true })
    })

    describe('createEvmNodeServiceSpec()', () => {
        it('should work when there are no anvil options', async () => {
            const spec = serializeDockerComposeSpec({
                version: '3.9',
                services: {
                    anvil: createEvmNodeServiceSpec({}),
                },
            })

            await writeFile(SPEC_FILE_PATH, spec)

            const result = validateSpec()

            expect(result.stderr.toString('utf8')).toBe('')
            expect(result.status).toBe(0)
            expect(spec).toMatchSnapshot()
        })

        it('should work with anvil options', async () => {
            await fc.assert(
                fc.asyncProperty(anvilOptionsArbitrary, async (anvilOptions) => {
                    const spec = serializeDockerComposeSpec({
                        version: '3.9',
                        services: {
                            anvil: createEvmNodeServiceSpec(anvilOptions),
                        },
                    })

                    await writeFile(SPEC_FILE_PATH, spec)

                    const result = validateSpec()

                    expect(result.stderr.toString('utf8')).toBe('')
                    expect(result.status).toBe(0)
                }),
                { numRuns: 20 }
            )
        })
    })

    describe('createEvmNodeProxyServiceSpec()', () => {
        const servicesArbitrary = fc.dictionary(
            serviceNameArbitrary,
            anvilOptionsArbitrary.map(createEvmNodeServiceSpec)
        )

        it('should work with anvil services', async () => {
            await fc.assert(
                fc.asyncProperty(portArbitrary, servicesArbitrary, async (port, services) => {
                    const spec = serializeDockerComposeSpec({
                        version: '3.9',
                        services: {
                            ...services,
                            rpc: createEvmNodeProxyServiceSpec(port, services),
                        },
                    })

                    await writeFile(SPEC_FILE_PATH, spec)

                    const result = validateSpec()

                    expect(result.stderr.toString('utf8')).toBe('')
                    expect(result.status).toBe(0)
                }),
                { numRuns: 20 }
            )
        })
    })

    describe('createSimulationComposeSpec()', () => {
        const anvilOptionsRecordArbitrary = fc.dictionary(serviceNameArbitrary, anvilOptionsArbitrary)

        it('should work goddammit', async () => {
            await fc.assert(
                fc.asyncProperty(portArbitrary, anvilOptionsRecordArbitrary, async (port, anvilOptions) => {
                    const simulationConfig = resolveSimulationConfig({ port }, hre.config)
                    const spec = serializeDockerComposeSpec(createSimulationComposeSpec(simulationConfig, anvilOptions))

                    await writeFile(SPEC_FILE_PATH, spec)

                    const result = validateSpec()

                    expect(result.stderr.toString('utf8')).toBe('')
                    expect(result.status).toBe(0)
                }),
                { numRuns: 20 }
            )
        })
    })
})
