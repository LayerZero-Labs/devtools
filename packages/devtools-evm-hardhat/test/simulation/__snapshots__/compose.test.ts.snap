// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simulation/compose createEvmNodeServiceSpec() should work when there are no anvil options 1`] = `
"version: '3.9'
services:
  anvil:
    build:
      dockerfile: Dockerfile
      target: node-evm
    command:
      - anvil
"
`;
